numba_dpex/core/utils/reduction_kernel_builder.py

```diff
@@ -0,0 +1,388 @@
+    redvars_legal_dict = reductionKernelVar.redvars_legal_dict
+
+    loc = parfor_node.init_block.loc
+
+    for race in parfor_node.races:
+        msg = (
+            "Variable %s used in parallel loop may be written "
+            "to simultaneously by multiple workers and may result "
+            "in non-deterministic or unintended results." % race
+        )
+        warnings.warn(NumbaParallelSafetyWarning(msg, loc))
+
+    loop_body_var_table = get_name_var_table(loop_body)
+    sentinel_name = get_unused_var_name("__sentinel__", loop_body_var_table)
+
+    global_size_var_name = []
+    global_size_mod_var_name = []
+    partial_sum_var_name = []
+    partial_sum_size_var_name = []
+    final_sum_var_name = []
+
+    for i, _ in enumerate(parfor_redvars):
+        reductionHelper = reductionHelperList[i]
+        name = reductionHelper.partial_sum_var.name
+        partial_sum_var_name.append(name)
+
+        name = reductionHelper.global_size_var.name
+        global_size_var_name.append(name)
+
+        name = reductionHelper.global_size_mod_var.name
+        global_size_mod_var_name.append(name)
+
+        name = reductionHelper.partial_sum_size_var.name
+        partial_sum_size_var_name.append(name)
+
+        name = reductionHelper.final_sum_var.name
+        final_sum_var_name.append(name)
+
+    kernel_name = "__numba_parfor_gufunc_%s_sum2" % (parfor_node.id)
+
+    partial_sum_var_dict = legalize_names(partial_sum_var_name)
+    global_size_var_dict = legalize_names(global_size_var_name)
+    global_size_mod_var_dict = legalize_names(global_size_mod_var_name)
+    partial_sum_size_var_dict = legalize_names(partial_sum_size_var_name)
+    final_sum_var_dict = legalize_names(final_sum_var_name)
+
+    partial_sum_var_legal_name = [
+        partial_sum_var_dict[v] for v in partial_sum_var_dict
+    ]
+
+    global_size_var_legal_name = [
+        global_size_var_dict[v] for v in global_size_var_dict
+    ]
+    global_size_mod_var_legal_name = [
+        global_size_mod_var_dict[v] for v in global_size_mod_var_dict
+    ]
+    partial_sum_size_var_legal_name = [
+        partial_sum_size_var_dict[v] for v in partial_sum_size_var_dict
+    ]
+    final_sum_var_legal_name = [
+        final_sum_var_dict[v] for v in final_sum_var_dict
+    ]
+
+    kernel_template = RemainderReduceIntermediateKernelTemplate(
+        kernel_name=kernel_name,
+        kernel_params=parfor_legalized_params,
+        sentinel_name=sentinel_name,
+        legal_loop_indices=legal_loop_indices,
+        redvars_dict=redvars_legal_dict,
+        redvars=parfor_redvars,
+        parfor_reddict=parfor_reddict,
+        typemap=typemap,
+        global_size_var_name=global_size_var_legal_name,
+        global_size_mod_var_name=global_size_mod_var_legal_name,
+        partial_sum_size_var_name=partial_sum_size_var_legal_name,
+        partial_sum_var_name=partial_sum_var_legal_name,
+        final_sum_var_name=final_sum_var_legal_name,
```

Comment:
For later, the number of arguments taken by the constructor of the template is too large. We should think of a cleaner design.